File: tree/Traversal/Postorder/recursive/cpp/PostorderTraversal.cpp
#include "PostorderTraversal.hpp"

#include <functional>
#include <vector>

using namespace std;

vector<int> PostorderTraversal(const vector<int>& leftChildren, const vector<int>& rightChildren, int root) {
    vector<int> order;

    function<void(int)> traverse = [&](int node) {
        if (leftChildren[node] != -1) {
            traverse(leftChildren[node]);
        }

        if (rightChildren[node] != -1) {
            traverse(rightChildren[node]);
        }

        order.push_back(node);
    };

    traverse(root);
    return order;
}

File: tree/Traversal/Postorder/recursive/java/PostorderTraversal.java
import java.util.*;

public class PostorderTraversal {
    public static List<Integer> PostorderTraversal(int[] leftChildren, int[] rightChildren, int root) {
        List<Integer> order = new ArrayList<>();
        traverse(root, leftChildren, rightChildren, order);
        return order;
    }

    private static void traverse(int node, int[] leftChildren, int[] rightChildren, List<Integer> order) {
        int leftChild = leftChildren[node];
        if (leftChild != -1) {
            traverse(leftChild, leftChildren, rightChildren, order);
        }

        int rightChild = rightChildren[node];
        if (rightChild != -1) {
            traverse(rightChild, leftChildren, rightChildren, order);
        }

        order.add(node);
    }
}
